Feat: new component multiselect [WIP] (#2188)
contentful/forma-36: packages/components/multiselect/README.mdx

---
title: 'Multiselect'
slug: /components/multiselect/
github: 'https://github.com/contentful/forma-36/tree/main/packages/components/multiselect'
storybook: 'https://v4-f36-storybook.netlify.app/?path=/story/components-multiselect--basic'
typescript: ./src/Multiselect.tsx
---

Search through a long list of options.

## Import

```jsx static=true
import { Multiselect } from '@contentful/f36-components';
// or
import { Multiselect } from '@contentful/f36-multiselect';
```

## Examples

### Basic usage

> The Multiselect requires 3 props to work:

Reviewer: ```suggestion
The Multiselect requires three props to work:
```

- `items`: It’s an array containing the items that will be shown as selectable options when the user types something in the `TextInput`.

> - `onInputValueChange`: This function will be called every time the user types something in the input.

Reviewer: Since this is a public interface, I would abstract some internal knowledge away to make it more smooth/ easier to read.
For the naming, my proposal would be `onSearchValueChange` since the users types something to search for stuff - while input is a very general name. The MultiSelect itself is an input.

  The component will pass the `item`, which the filter method is currently iterating over, and the `inputValue` prop of the `TextInput` component.
- `onSelectItem`: This function is called when the user selects one of the options of the list. The component will pass the selected item as an argument to the function.

> An `Multiselect` with a list of spaces will look like this:

Reviewer: ```suggestion
A `Multiselect` with a list of spaces will look like this:
```

```jsx file=./examples/MultiselectBasicUsageExample.tsx

```

### Using objects as `items`

> We showed how to create an Multiselect with an array of string but it’s also possible to use other types of data as `items`.
> A very common way of using the Multiselect is with objects and for that, with a few changes to the previous example this can be done:

Reviewer: ```suggestion
We showed how to create a Multiselect with an array of strings but it’s also possible to use other data types as `items`.
A widespread way of using Multiselect is with objects and for that, with a few changes to the previous example this can be done:
```

_Suggestions by Grammarly_

```jsx file=./examples/MultiselectWithObjectsExample.tsx

```

Both `itemToString` and `renderItem` are necessary when passing objects as items and they both will receive an "item" as an argument.

If you are using Typescript, you can tell the Multiselect what is the type of your items to make these functions strongly typed.
You can do that by writing the component like this `<Multiselect<ItemType> {...props}/>`

> ### Highlighting an item with `getStringMatch`

Reviewer: Looks like the rest of the documentation isn't updated at all.

Author: yes, still WIP :)

---
title: 'Multiselect'
type: 'component'
status: 'alpha'
section: 'formComponents'
slug: /components/multiselect/
github: 'https://github.com/contentful/forma-36/tree/main/packages/components/forms/src/Multiselect'
storybook: 'https://v4-f36-storybook.netlify.app/?path=/story/form-elements-multiselect--basic'
typescript: ./Multiselect.tsx,./MultiselectOption
---

Multiselect allows users to make a multi selection from a list of options.
It has multiple optional features like a search.

## Import

```jsx static=true
import { Multiselect } from '@contentful/f36-components';
// or
import { Select } from '@contentful/f36-forms';
```

## Examples

### Basic usage

> The Multiselect is a fully controlled from element and expects the selectable items as children via the compund component Multiselect.Option.

Reviewer: ```suggestion
The Multiselect is a fully controlled form element and expects the selectable items as children via the compound component Multiselect.Option.
```

Properties are:

> for the multiselect itself:

Reviewer: ```suggestion
For the multiselect itself:
```

- label: displayed label
- onSelectItem: This function is called when the user selects a option, it needs to be passed to the MultiselectOption
- isChecked (optional): controlls if a element is initially selected
- isDisabled (optional): controlls if a element is changeable

```jsx file=./examples/MultiselectBasicUsageExample.tsx

```

### Searchable options

One optional feature of the Multiselect component is to allow filtering the list via a search field.
To make it work, the onSearchValueChange callback function needs to be provided.

To enable searching the following properties have to be set.

- onSearchValueChange: Callbackfunction which enables the search field. This function needs to provide the search /filter algorithm
- searchPlaceholder (optional): placeholder for the search field
- noMatchesMessage (optional): message shown when search result is empty

```jsx file=./examples/MultiselectBasicUsageExample.tsx

```

## Props (API reference)

<PropsTable of="Multiselect" />

## Content guidelines

- Multiselect label should be short, contain 1 to 3 words
- Label should be written in a sentence case (the first word capitalized, the rest lowercase)

## Accessibility

> - dismisses the dropdown when selecting with the enter key

Reviewer: ```suggestion
- Dismisses the dropdown when selecting with the enter key
```

Properties are:

for the multiselect itself:

- currentSelection (optional but recommendet): an array of the labels of selected options. This will give the user a hint that there are selected items. This is optional
- placeholder (optional): Label of the trigger button without selected elements
- startIcon (optional): Icon on the start of the trigger button
- listWidth (optional): auto or full
- listMaxHeight (optional): sets the maximum height of the list drawer, overflow will be scrolled
- isLoading: enables showing a loading animation while searching or loading results

It also accepts refs for the toggle button, the searchInput and the list.

For the option:

- itemId: identifier for the element, used to link the label with the checkbox
- value: the actual value of the select option
- label: displayed label
- onSelectItem: This function is called when the user selects a option, it needs to be passed to the MultiselectOption
- isChecked (optional): controlls if a element is initially selected
- isDisabled (optional): controlls if a element is changeable

```jsx file=./examples/MultiselectBasicUsageExample.tsx

```

### Searchable options

> One of optional feature of the Multiselect component is to allow filtering the list via a search field.

Reviewer: ```suggestion
One optional feature of the Multiselect component is to allow filtering the list via a search field.
```

---
title: 'Multiselect'
type: 'component'
status: 'alpha'
section: 'formComponents'
slug: /components/multiselect/
github: 'https://github.com/contentful/forma-36/tree/main/packages/components/forms/src/Multiselect'
storybook: 'https://v4-f36-storybook.netlify.app/?path=/story/form-elements-multiselect--basic'
typescript: ./Multiselect.tsx,./MultiselectOption
---

Multiselect allows users to make a multi selection from a list of options.
It has multiple optional features like a search.

## Import

```jsx static=true
import { Multiselect } from '@contentful/f36-components';
// or

> import { Select } from '@contentful/f36-forms';

Reviewer 2: ```suggestion
import { Multiselect } from '@contentful/f36-forms';
```

Author: ahh another one slipped trough. will leave this out for now completely until it is a stable part of the forms package.
